Build failure: pkgsLLVM.tbb

### Steps To Reproduce

Steps to reproduce the behavior:
1. `nix build github:NixOS/nixpkgs#pkgsLLVM.tbb`

### Build log

```
error: builder for '/nix/store/a0kifxqrpgjbnzaxf9c8vgc038spjsfr-tbb-x86_64-unknown-linux-gnu-2020.3.drv' failed with exit code 2;
       last 25 log lines:
       > x86_64-unknown-linux-gnu-ld: error: version script assignment of 'local' to symbol 'set_memcpy_largest_cachelinesize' failed: symbol not defined
       > x86_64-unknown-linux-gnu-ld: error: version script assignment of 'local' to symbol 'set_memcpy_largest_cache_size' failed: symbol not defined
       > clang: error: linker command failed with exit code 1 (use -v to see invocation)
       > make[1]: *** [../../build/Makefile.tbbmalloc:85: libtbbmalloc.so.2] Error 1
       > make[1]: Leaving directory '/build/source/build/linux_intel64_clang_cc_libc2.39_kernel6.1.0_release'
       > make: *** [Makefile:32: tbbmalloc] Error 2
       > make: *** Waiting for unfinished jobs....
       > 1 warning generated.
       > x86_64-unknown-linux-gnu-clang++ -fPIC -o libtbb.so.2 concurrent_hash_map.o concurrent_queue.o concurrent_vector.o dynamic_link.o itt_notify.o cache_aligned_allocator.o pipeline.o queuing_mutex.o queuing_rw_mutex.o reader_writer_lock.o spin_rw_mutex.o x86_rtm_rw_mutex.o spin_mutex.o critical_section.o mutex.o recursive_mutex.o condition_variable.o tbb_thread.o concurrent_monitor.o semaphore.o private_server.o rml_tbb.o tbb_misc.o tbb_misc_ex.o task.o task_group_context.o governor.o market.o arena.o scheduler.o observer_proxy.o tbb_statistics.o tbb_main.o concurrent_vector_v2.o concurrent_queue_v2.o spin_rw_mutex_v2.o task_v2.o   -ldl -lrt -shared -Wl,-soname=libtbb.so.2 -pthread -m64  -Wl,--version-script,tbb.def
       > x86_64-unknown-linux-gnu-ld: error: version script assignment of 'local' to symbol 'get_msg_buf' failed: symbol not defined
       > x86_64-unknown-linux-gnu-ld: error: version script assignment of 'local' to symbol 'get_text_buf' failed: symbol not defined
       > x86_64-unknown-linux-gnu-ld: error: version script assignment of 'local' to symbol 'message_catalog' failed: symbol not defined
       > x86_64-unknown-linux-gnu-ld: error: version script assignment of 'local' to symbol 'print_buf' failed: symbol not defined
       > x86_64-unknown-linux-gnu-ld: error: version script assignment of 'local' to symbol 'irc__get_msg' failed: symbol not defined
       > x86_64-unknown-linux-gnu-ld: error: version script assignment of 'local' to symbol 'irc__print' failed: symbol not defined
       > x86_64-unknown-linux-gnu-ld: error: version script assignment of 'global' to symbol '_ZN3tbb10empty_taskD1Ev' failed: symbol not defined
       > x86_64-unknown-linux-gnu-ld: error: version script assignment of 'global' to symbol '_ZN3tbb14bad_last_allocD1Ev' failed: symbol not defined
       > x86_64-unknown-linux-gnu-ld: error: version script assignment of 'global' to symbol '_ZN3tbb12missing_waitD1Ev' failed: symbol not defined
       > x86_64-unknown-linux-gnu-ld: error: version script assignment of 'global' to symbol '_ZN3tbb27invalid_multiple_schedulingD1Ev' failed: symbol not defined
       > x86_64-unknown-linux-gnu-ld: error: version script assignment of 'global' to symbol '_ZN3tbb13improper_lockD1Ev' failed: symbol not defined
       > x86_64-unknown-linux-gnu-ld: error: version script assignment of 'global' to symbol '_ZN3tbb10user_abortD1Ev' failed: symbol not defined
       > clang++: error: linker command failed with exit code 1 (use -v to see invocation)
       > make[1]: *** [../../build/Makefile.tbb:103: libtbb.so.2] Error 1
       > make[1]: Leaving directory '/build/source/build/linux_intel64_clang_cc_libc2.39_kernel6.1.0_release'
       > make: *** [Makefile:29: tbb] Error 2
       For full logs, run 'nix log /nix/store/a0kifxqrpgjbnzaxf9c8vgc038spjsfr-tbb-x86_64-unknown-linux-gnu-2020.3.drv'.

```

### Additional context

As you can see, using pkgsLLVM when building tbb causes issues due to LLD >=17.

### Notify maintainers

<!--
Please @ people who are in the `meta.maintainers` list of the offending package or module.
If in doubt, check `git blame` for whoever last touched something.
-->

### Metadata

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"

```
       error: file 'nixpkgs' was not found in the Nix search path (add it using $NIX_PATH or -I)
---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/development/libraries/tbb/2020_3.nix
+++ b/pkgs/development/libraries/tbb/2020_3.nix
@@ -60,6 +60,9 @@ stdenv.mkDerivation rec {
     else if stdenv.hostPlatform.isi686 then "arch=ia32"
     else throw "Unsupported cross architecture"));
 
+  # Fix undefined reference errors with version script under LLVM.
+  NIX_LDFLAGS = lib.optionalString (stdenv.cc.bintools.isLLVM && lib.versionAtLeast stdenv.cc.bintools.version "17") "--undefined-version";
+
   enableParallelBuilding = true;
 
   installPhase = ''

--- a/pkgs/development/libraries/tbb/default.nix
+++ b/pkgs/development/libraries/tbb/default.nix
@@ -39,6 +39,9 @@ stdenv.mkDerivation rec {
     # https://gcc.gnu.org/PR108854
     lib.optionals (stdenv.cc.isGNU && stdenv.isx86_32) [ "-O2" ];
 
+  # Fix undefined reference errors with version script under LLVM.
+  NIX_LDFLAGS = lib.optionalString (stdenv.cc.bintools.isLLVM && lib.versionAtLeast stdenv.cc.bintools.version "17") "--undefined-version";
+
   # Disable failing test on musl
   # test/conformance/conformance_resumable_tasks.cpp:37:24: error: ‘suspend’ is not a member of ‘tbb::v1::task’; did you mean ‘tbb::detail::r1::suspend’?
   postPatch = lib.optionalString stdenv.hostPlatform.isMusl ''

